File: lab-grpc/java/grpc-central/src/main/java/br/pucminas/labdamd/central/ServidorCentral.java
package br.pucminas.labdamd.central;

import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.stub.StreamObserver;

import java.io.IOException;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;

// Parte C/D - servidor gRPC da Central. Repare que não abro socket nem trato
// bytes: eu só implemento os métodos do serviço que foram definidos no
// central.proto. O gRPC cuida da rede (HTTP/2), da serialização, etc.
public class ServidorCentral {
    // OFFSET = 2 últimos dígitos do RA (ver README). Cliente usa o mesmo.
    static final int OFFSET = 39;

    public static void main(String[] args) throws IOException, InterruptedException {
        int porta = 50051 + OFFSET; // 50090
        Server servidor = ServerBuilder.forPort(porta)
                .addService(new CentralAtendimentoImpl())
                .build();
        servidor.start();
        System.out.println("[gRPC] Servidor da Central ouvindo na porta " + porta);
        servidor.awaitTermination();
    }

    // Implementação do serviço: estende a classe-base gerada a partir do .proto.
    static class CentralAtendimentoImpl extends CentralAtendimentoGrpc.CentralAtendimentoImplBase {

        // RPC unário: recebe uma PerguntaHorario, devolve uma RespostaHorario.
        @Override
        public void consultarHorario(PerguntaHorario pedido, StreamObserver<RespostaHorario> observador) {
            String horario = LocalTime.now().format(DateTimeFormatter.ofPattern("HH:mm:ss"));
            System.out.println("[gRPC] ConsultarHorario chamado por: " + pedido.getNomeAluno());
            RespostaHorario resposta = RespostaHorario.newBuilder()
                    .setHorarioAtual(horario)
                    .setMensagem("Olá, " + pedido.getNomeAluno() + "! Agora são " + horario + ".")
                    .build();
            // onNext envia a resposta; onCompleted encerra a chamada (é unária: uma só).
            observador.onNext(resposta);
            observador.onCompleted();
        }
    }
}

File: lab-grpc/python/grpc_central/servidor_central.py
from concurrent import futures
from datetime import datetime
import time

import grpc
import central_pb2
import central_pb2_grpc

# Parte C/D - servidor gRPC em Python. Igual ao Java: eu só implemento os
# métodos do serviço; o gRPC cuida da rede e da serialização.
# OFFSET = 2 últimos dígitos do RA (ver README).
OFFSET = 39
PORTA = 50061 + OFFSET  # 50100


class CentralAtendimentoServicer(central_pb2_grpc.CentralAtendimentoServicer):

    # RPC unário: recebe uma PerguntaHorario, devolve uma RespostaHorario.
    def ConsultarHorario(self, request, context):
        horario = datetime.now().strftime("%H:%M:%S")
        print(f"[gRPC] ConsultarHorario chamado por: {request.nome_aluno}")
        return central_pb2.RespostaHorario(
            horario_atual=horario,
            mensagem=f"Olá, {request.nome_aluno}! Agora são {horario}.",
        )

    # RPC com streaming de servidor. Em Python é só uma função geradora: cada
    # 'yield' envia mais um Aviso ao cliente, na mesma conexão.
    def AcompanharAvisos(self, request, context):
        print(f"[gRPC] AcompanharAvisos: {request.nome_aluno} se inscreveu.")
        for i in range(1, 6):
            yield central_pb2.Aviso(
                numero=i,
                texto=f"Aviso #{i}: a aula começa em {5 - i} minuto(s)!",
            )
            time.sleep(2)


def main():
    servidor = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    central_pb2_grpc.add_CentralAtendimentoServicer_to_server(CentralAtendimentoServicer(), servidor)
    servidor.add_insecure_port(f"[::]:{PORTA}")
    servidor.start()
    print(f"[gRPC] Servidor da Central ouvindo na porta {PORTA}")
    servidor.wait_for_termination()


if __name__ == "__main__":
    main()
